File: sections/18_dsu_minimum_spanning_trees/problems/a_network_growth/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct DSU {
    vector<int> parent, sz;
    DSU(int n) : parent(n), sz(n, 1) {
        iota(parent.begin(), parent.end(), 0);
    }
    int find(int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }
    int unite(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return sz[ra];
        if (sz[ra] < sz[rb]) swap(ra, rb);
        parent[rb] = ra;
        sz[ra] += sz[rb];
        return sz[ra];
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    if (!(cin >> n >> m)) return 0;
    DSU dsu(n);
    int components = n, largest = 1;
    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        --u;
        --v;
        if (dsu.find(u) != dsu.find(v)) {
            largest = max(largest, dsu.unite(u, v));
            components--;
        }
        cout << components << ' ' << largest << '\n';
    }
    return 0;
}

File: sections/18_dsu_minimum_spanning_trees/problems/a_network_growth/solution.py
import sys


class DSU:
    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.size = [1] * n

    def find(self, x: int) -> int:
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

    def union(self, a: int, b: int) -> int:
        ra = self.find(a)
        rb = self.find(b)
        if ra == rb:
            return self.size[ra]
        if self.size[ra] < self.size[rb]:
            ra, rb = rb, ra
        self.parent[rb] = ra
        self.size[ra] += self.size[rb]
        return self.size[ra]


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    it = iter(data)
    n = next(it)
    m = next(it)
    dsu = DSU(n)
    components = n
    largest = 1
    out: list[str] = []
    for _ in range(m):
        u = next(it) - 1
        v = next(it) - 1
        ru = dsu.find(u)
        rv = dsu.find(v)
        if ru != rv:
            merged = dsu.union(ru, rv)
            components -= 1
            largest = max(largest, merged)
        out.append(f"{components} {largest}")
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

